src/vehicle_autodrive.cpp

```diff
@@ -1,25 +1,1202 @@
+ * is obtained from driver.omt_path). Thus we always plan ahead 2 OMTs worth of navigation,
+ * but we only follow the first half of the route, then recompute a new route for the next
+ * 2 OMTs. This ensures that we don't drive up to the edge of an OMT only to get stuck due
+ * to a large obstacle that can't be driven around from that position (a dead end) and instead
+ * we always choose to cross OMT boundaries in a position from which we are guaranteed to be
+ * able to continue onward.
+ *
+ * The pathfinding algorithm being employed is A*. The graph being searched is made up of nodes
+ * denoted by x and y coordinates as well as a direction of travel. Since steering changes the
+ * vehicle's direction in increments of 15 degrees there are 360/15 possible orientations. By
+ * limiting navigation to 2 OMTs at a time we have a maximum search space of 2*24*24*24 nodes.
+ * In practice, A* finds a good path after visiting fewer than 50 nodes due to the heuristic
+ * function that estimates which nodes are most likely to lead to the destination (which is
+ * based on proximity to the "exit zone" leading to the next OMT and whether the vehicle is
+ * pointing towards the middle of that zone). Although there can be many paths that we could
+ * take to get through to the next OMT, we just take the first one we find since it tends to
+ * be the most direct and is usually near-optimal. When the way is blocked we'll either give
+ * up after exhausting the search space or after we hit a limit on the search length.
+ *
+ * In order to build the search graph for A*, we consider every possible map tile and orientation
+ * in the search zone and we figure out whether the vehicle's pivot point could be placed there
+ * without the vehicle overlapping some obstacle (a collision-free placement). The pivot point
+ * is the part of the vehicle that moves the least while steering and is usually between the
+ * wheels of the vehicle, closer to the non-steerable ones. The pathfinding problem reduces to
+ * finding a path for the pivot point to follow through the set of possible positions and
+ * orientations (nodes). In order to figure out the possible connections between nodes (the
+ * edges of the graph) we consider a limited set of actions that the driver can perform at
+ * each location: nothing, steer left or steer right (possibly more than once). Given a current
+ * node, current speed and amount of steering we can predict where the vehicle's pivot point
+ * will end up in 1 turn and connect the 2 nodes.
+ *
+ * In order to keep the search space small we only consider paths traveled at constant speed
+ * (otherwise we'd have to add speed as an extra attribute in every node). When we need to
+ * compute a new path we first try to use the maximum safe autodrive speed (which is typically
+ * much less than the vehicle's max speed). If we can't find a path we cut the speed in half and
```

Comment:
For the future (I definitely agree not to hold up the patch until then!), what are the main things to work on to enable a faster autodrive speed?

Reply:
Speed can be increased when we're able to reliably steer more than once per turn. The main blocker for that is accurately estimating the move cost of steering. Unfortunately that depends on the terrain that the wheels are on at the time of the turn and I'd need to set up a new precomputed data structure for that.

Reply:
But the player might not want to increase the autodrive speed too much because:
a) it burns more fuel (maybe not for helicopters)
b) in case of crash, the damage is higher
The plan is to give the player control over this eventually so they can use their own risk tolerance.